File: DSA/src/Questions/StackQueue/QueueUsingStack.java
package Questions.StackQueue;

import java.util.Stack;

/*
https://leetcode.com/problems/implement-queue-using-stacks/description/
 Time Complexity:
Push: O(n) where n is the number of elements in the queue. This is because we move elements between the two stacks.
Pop: O(1) for removing the top element from s1.
Peek: O(1) for getting the top element from s1.
Empty: O(1) for checking if s1 is empty.
Space Complexity: O(n) for storing the elements in the two stacks.
 */
public class QueueUsingStack {
    private final Stack<Integer> s1 = new Stack<>();
    private final Stack<Integer> s2 = new Stack<>();

    public QueueUsingStack() {
    }

    public void push(int x) {
        while (!s1.isEmpty()) {
            s2.push(s1.pop());
        }
        s1.push(x);
        while (!s2.isEmpty()) {
            s1.push(s2.pop());
        }
    }

    public int pop() {
        return s1.pop();
    }

    public int peek() {
        return s1.peek();
    }

    public boolean empty() {
        return s1.isEmpty();
    }
    public static void main(String[] args) {
        QueueUsingStack queue = new QueueUsingStack();
        queue.push(1);
        queue.push(2);
        queue.push(3);
        queue.push(4);
        queue.push(5);
        System.out.println("Top element of the queue: " + queue.peek());
    }
}

File: DSA/Python/Questions/StackQueue/QueueUsingStack.py
"""
# Question: Implement Queue using Stacks
# Link: https://leetcode.com/problems/implement-queue-using-stacks/

# Implement queue operations using two stacks

# Time Complexity:
# - Push: O(n)
# - Pop: O(1)
# - Peek: O(1)
# - Empty: O(1)

# Space Complexity: O(n)

# Algorithm:
# 1. Use two stacks for queue simulation
# 2. Transfer elements between stacks for push
# 3. Maintain FIFO order
# 4. Optimize pop and peek operations

# Key Components:
# - MyQueue class implementation
# - Two stacks for storage
# - Element transfer logic
"""


class MyQueue:
    def __init__(self):
        self.s1 = []  # main stack
        self.s2 = []  # auxiliary stack

    def push(self, x: int) -> None:
        # Transfer all elements from s1 to s2
        while self.s1:
            self.s2.append(self.s1.pop())

        # Add new element to s1
        self.s1.append(x)

        # Transfer back all elements from s2 to s1
        while self.s2:
            self.s1.append(self.s2.pop())

    def pop(self) -> int:
        return self.s1.pop()

    def peek(self) -> int:
        return self.s1[-1]

    def empty(self) -> bool:
        return len(self.s1) == 0


def main():
    queue = MyQueue()
    queue.push(1)
    queue.push(2)
    queue.push(3)
    print(queue.peek())  # Expected: 1
    print(queue.pop())  # Expected: 1
    print(queue.empty())  # Expected: False


if __name__ == "__main__":
    main()
